### PR TITLE
[BanBye] Add extractor

### LINKED ISSUE #3175: Banbye support
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.03.08.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

Europe

### Example URLs

https://banbye.com/watch/v_ytfmvkVYLE8T

### Description

New popular video site, unsupported yet

### Verbose log

```shell
./yt-dlp https://banbye.com/watch/v_ytfmvkVYLE8T
[generic] v_ytfmvkVYLE8T: Requesting header
[redirect] Following redirect to https://banbye.com/en/watch/v_ytfmvkVYLE8T
[generic] v_ytfmvkVYLE8T: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] v_ytfmvkVYLE8T: Downloading webpage
[generic] v_ytfmvkVYLE8T: Extracting information
ERROR: Unsupported URL: https://banbye.com/en/watch/v_ytfmvkVYLE8T
```

### PR DESCRIPTION
**Pull request by mehq**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Closes #3175.

### CODE REVIEW

File: yt_dlp/extractor/banbye.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class BanByeBaseIE(InfoExtractor):
+    _API_BASE = 'https://api.banbye.com'
+    _CDN_BASE = 'https://cdn.banbye.com'
+
+    def _extract_playlist(self, playlist_id):
+        data = self._download_json('%s/playlists/%s' % (self._API_BASE, playlist_id), playlist_id)
+        return self.playlist_result([
+            self.url_result('https://banbye.com/watch/%s' % video_id, ie_key='BanBye')
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:16]:
```suggestion
            self.url_result(f'https://banbye.com/watch/{video_id}', BanByeIE)
```

File: yt_dlp/extractor/banbye.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            for video_id in data['videoIds']], playlist_id, data['name'])
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:17]:
```suggestion
            for video_id in data['videoIds']], playlist_id, data.get('name'))
```

File: yt_dlp/extractor/banbye.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+
+class BanByeIE(BanByeBaseIE):
+    _VALID_URL = r'''(?x)
+                    https?://
+                        (?:www\.)?banbye.com/
+                        (?:en/)?watch/
+                        (?P<id>[^/?\#&]+)/?
+                        \??(?:playlistId=(?P<playlist_id>[^/?\#&]+))?'''
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:26]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?banbye.com/(?:en/)?watch/(?P<id>\w+)'
```
there may be other queries in the URL. Better to take playlist_id from parse_url

File: yt_dlp/extractor/banbye.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        mobj = self._match_valid_url(url)
+        video_id = mobj.group('id')
+        playlist_id = mobj.group('playlist_id')
+
+        if playlist_id and not self.get_param('noplaylist'):
+            self.to_screen(f'Downloading playlist {playlist_id}; add --no-playlist to just download video {video_id}')
+            return self._extract_playlist(playlist_id)
+
+        if playlist_id:
+            self.to_screen(f'Downloading just video {video_id} because of --no-playlist')
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:67]: use `self._yes_playlist`

File: yt_dlp/extractor/banbye.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            'channel_url': 'https://banbye.com/channel/%s' % data['channelId'],
+            'timestamp': unified_timestamp(data['publishedAt']),
+            'duration': data['duration'],
+            'tags': data['tags'],
+            'formats': formats,
+            'thumbnails': thumbnails,
+            'like_count': data['likes'],
+            'dislike_count': data['dislikes'],
+            'view_count': data['views'],
+            'comment_count': data['commentCount'],
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:97]: make all of this non-fatal

File: yt_dlp/extractor/banbye.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+
+class BanByePlaylistIE(BanByeBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?banbye.com/(?:en/)?channel/[^/]+/?\?playlist=(?P<id>[^/?\#&]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:102]: Same issue as the other one. URLs like:
```
https://banbye.com/channel/ch_wrealu24?a=1&playlist=p_Ld82N6gBw_OJ
```
should match

**Review by pukkandan — CHANGES_REQUESTED** [review]
you may also want to add channel support if possible

File: yt_dlp/extractor/banbye.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+class BanByeBaseIE(InfoExtractor):
+    _API_BASE = 'https://api.banbye.com'
+    _CDN_BASE = 'https://cdn.banbye.com'
+    _VIDEO_BASE = 'https://banbye.com/watch'
+
+    @staticmethod
+    def _extract_playlist_id(url, param='playlist'):
+        return compat_parse_qs(
+            compat_urllib_parse_urlparse(url).query).get(param, [None])[0]
```
Comment (Ashish0804) [inline yt_dlp/extractor/banbye.py:26]:
```suggestion
        return (compat_parse_qs(
            compat_urllib_parse_urlparse(url).query).get(param) or [None])[0]
```

File: yt_dlp/extractor/banbye.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    def _extract_playlist(self, playlist_id):
+        data = self._download_json(f'{self._API_BASE}/playlists/{playlist_id}', playlist_id)
+        return self.playlist_result([
+            self.url_result(f'{self._VIDEO_BASE}/{video_id}', BanByeIE)
+            for video_id in data['videoIds']], playlist_id, data.get('name'))
```
Comment (Ashish0804) [inline yt_dlp/extractor/banbye.py:32]: this shouldn't be fatal imo
```suggestion
            for video_id in data.get('videoIds') or []], playlist_id, data.get('name'))
```

File: yt_dlp/extractor/banbye.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        data = self._download_json(f'{self._API_BASE}/videos/{video_id}', video_id)
+        thumbnails = [{
+            'id': f'{quality}p',
+            'url': f'{self._CDN_BASE}/video/{video_id}/{quality}.webp',
+        } for quality in [48, 96, 144, 240, 512, 1080]]
+        formats = [{
+            'format_id': f'http-{quality}p',
+            'quality': quality,
+            'url': f'{self._CDN_BASE}/video/{video_id}/{quality}.mp4',
+        } for quality in data['quality']]
```
Comment (Ashish0804) [inline yt_dlp/extractor/banbye.py:83]:
```suggestion
        } for quality in data.get('quality') or []]
```

File: yt_dlp/extractor/banbye.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': data.get('title'),
+            'description': data.get('desc'),
+            'uploader': data.get('channel', {}).get('name'),
```
Comment (Ashish0804) [inline yt_dlp/extractor/banbye.py:91]:
```suggestion
            'uploader': (data.get('channel') or {}).get('name'),
```

File: yt_dlp/extractor/banbye.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+            'channel_url': format_field(data.get('channelId'), template='https://banbye.com/channel/%s'),
+            'timestamp': unified_timestamp(data.get('publishedAt')),
+            'duration': data.get('duration'),
+            'tags': data.get('tags'),
+            'formats': formats,
+            'thumbnails': thumbnails,
+            'like_count': data.get('likes'),
+            'dislike_count': data.get('dislikes'),
+            'view_count': data.get('views'),
+            'comment_count': data.get('commentCount'),
```
Comment (Ashish0804) [inline yt_dlp/extractor/banbye.py:102]: use int_or_none

File: yt_dlp/extractor/banbye.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        def page_func(page_num):
+            data = self._download_json(f'{self._API_BASE}/videos', channel_id, query={
+                'channelId': channel_id,
+                'sort': 'new',
+                'limit': self._PAGE_SIZE,
+                'offset': page_num * self._PAGE_SIZE,
+            }, note=f'Downloading page {page_num+1}')
+            return [
+                self.url_result(f"{self._VIDEO_BASE}/{video['_id']}", BanByeIE)
+                for video in data['items']
```
Comment (Ashish0804) [inline yt_dlp/extractor/banbye.py:142]:
```suggestion
                for video in data.get('items') or []
```

File: yt_dlp/extractor/banbye.py (hunk 8 above)
Comment (mehq) [reply yt_dlp/extractor/banbye.py:32]: why not? u cant make playlist data without `videoIds`. imo an error would be better than an empty playlist when the extraction process breaks

Comment (Ashish0804) [reply yt_dlp/extractor/banbye.py:32]: playlist extraction shouldn't break if videos are not avail imo. U just give a playlist with "0" entries and the core code then prints out no videos were found...that's much better

File: yt_dlp/extractor/banbye.py (hunk 10 above)
Comment (pukkandan) [reply yt_dlp/extractor/banbye.py:91]:
```suggestion
            'uploader': traverse_obj(data, ('channel' 'name')),
```

File: yt_dlp/extractor/banbye.py (hunk 8 above)
Comment (fstirlitz) [reply yt_dlp/extractor/banbye.py:32]: If the key is absent from the JSON, that means the extractor is no longer correct, and this is a bug that needs to be fxed. The program should not sweep that under the rug and pretend everything is fine.

I expect an empty playlist to return an empty list in `videoIds` anyway.

**Comment by Ashish0804** [comment]
@mehq u just marked most of the comments as resolved without even making any changes...
If my review is not even worth a reply then okay i won't waste time next time reviewing.

**Comment by mehq** [comment]
@Ashish0804 i did not intend to offend u but most of them were unnecessary

File: yt_dlp/extractor/banbye.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': data.get('title'),
+            'description': data.get('desc'),
+            'uploader': traverse_obj(data, ('channel', 'name')),
+            'channel_id': data.get('channelId'),
+            'channel_url': format_field(data.get('channelId'), template='https://banbye.com/channel/%s'),
```
Comment (pukkandan) [inline yt_dlp/extractor/banbye.py:94]:
```suggestion
            'channel_url': format_field(data, 'channelId', 'https://banbye.com/channel/%s'),
```

**Review by pukkandan — APPROVED** [review]

**Pull request merged**